### PR TITLE
Maintenance window support

### PR DESCRIPTION
**Pull request by nlamirault**
This PR add support for https://docs.microsoft.com/en-us/azure/aks/planned-maintenance

### CODE REVIEW

**Comment by nlamirault** [comment]
@jcorioland Any news for this feature ?

**Comment by jcorioland** [comment]
@nlamirault I am contributor on this project, but I can't review/merge PR, sorry.

**Comment by nlamirault** [comment]
ok @jcorioland .
Project seems not maintained ? Last commit is yours (2021/05/27)

**Comment by lonegunmanb** [comment]
> ok @jcorioland . Project seems not maintained ? Last commit is yours (2021/05/27)

Hi @nlamirault , @jcorioland was a contributor and his's pr was accepted and merged by the project's owner.

**Comment by nlamirault** [comment]
Could you review this PR @yupwei68 please ?

**Comment by lonegunmanb** [comment]
Hello @nlamirault , thanks for your pr! I'd like to merge this pr, but now we have conflicts to resolve, and I've left some comments, would you please take a look and push again? Thanks!

**Review by lonegunmanb — CHANGES_REQUESTED** [review]

File: main.tf [hunk 1]
```diff
@@ -84,6 +84,24 @@ resource "azurerm_kubernetes_cluster" "main" {
     }
   }
 
+  maintenance_window {
+    dynamic "allowed" {
+      for_each = var.enable_maintenance_window == true ? var.maintenance_allowed : []
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply main.tf:90]: Done

File: main.tf [hunk 2]
```diff
@@ -84,6 +84,24 @@ resource "azurerm_kubernetes_cluster" "main" {
...
+      content {
+        day = allowed.value.day
+        hours = allowed.value.hours
+      }
+    }
+
+    dynamic "not_allowed" {
+      for_each = var.enable_maintenance_window == true ? var.maintenance_not_allowed : []
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply main.tf:98]: Done

File: variables.tf [hunk 3]
```diff
@@ -289,3 +289,27 @@ variable "enable_host_encryption" {
...
+  description = "Enable maintenance for AKS cluster"
+  type        = bool
+  default     = false
+}
+
+variable "maintenance_allowed" {
+  description = "Days and hours when maintenance is allowed"
+  type = list(object({
+    day = string
+    hours = list(string)
```
[the remark this answers is not in the document]
Comment (nlamirault) [reply variables.tf]: Done

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Hi, I've tested the code with fixture in the test folder, and the following error occurred:

```txt
│ Error: Missing required argument
│
│   with module.aks.azurerm_kubernetes_cluster.main,
│   on ..\..\main.tf line 88, in resource "azurerm_kubernetes_cluster" "main":
│   88:   maintenance_window {
│
│ "maintenance_window.0.allowed": one of `maintenance_window.0.allowed,maintenance_window.0.not_allowed` must be specified
╵
╷
│ Error: Missing required argument
│
│   with module.aks.azurerm_kubernetes_cluster.main,
│   on ..\..\main.tf line 88, in resource "azurerm_kubernetes_cluster" "main":
│   88:   maintenance_window {
│
│ "maintenance_window.0.not_allowed": one of `maintenance_window.0.allowed,maintenance_window.0.not_allowed` must be specified
╵
```

The `enable_maintenance_window` default value `false` will emit an empty `maintenance_window` block which is not valid. Would the following code be better?:

```hcl
dynamic "maintenance_window" {
    for_each = var.enable_maintenance_window ? ["maintenance_window"] : []
    content {
      dynamic "allowed" {
        for_each = var.maintenance_allowed
        content {
          day   = allowed.value.day
          hours = allowed.value.hours
        }
      }

      dynamic "not_allowed" {
        for_each = var.maintenance_not_allowed
        content {
          start = not_allowed.value.start
          end   = not_allowed.value.end
        }
      }
    }
  }
```

**Comment by zioproto** [comment]
@nlamirault this PR does not merge anymore in the current master. Please rebase this work asap or look into the proposed #256

**Comment by lonegunmanb** [comment]
That's ok @zioproto, I'm working on a new pr for this feature, I'll submit it when it's ready.

Btw, in my new pr, I'd like to use [`optional`](https://www.terraform.io/language/expressions/type-constraints#optional-object-type-attributes) feature introduced by Terraform 1.3, which requires we upgrade Terraform Core's restriction to `>= 1.3`. Do you think such a change is a breaking change? If so, I'd like to defer this pr to our next new major release, maybe after we solve all the remaining prs.

**Comment by lonegunmanb** [comment]
Thanks @nlamirault for open this pr, I'm closing it in favor of #256 has brought us this feature. Thanks for your contribution to this module.

**Comment by nlamirault** [comment]
@lonegunmanb :+1:

**Pull request closed without merging**